src/librustc_mir/interpret/operator.rs

```diff
@@ -142,82 +145,16 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
-                }
-                Ne if left_kind == right_kind => {
-                    let result = match (left, right) {
-                        (PrimVal::Bytes(left), PrimVal::Bytes(right)) => left != right,
-                        (PrimVal::Ptr(left), PrimVal::Ptr(right)) => left != right,
-                        (PrimVal::Undef, _) | (_, PrimVal::Undef) => return Err(EvalError::ReadUndefBytes),
-                        _ => true,
-                    };
-                    return Ok((PrimVal::from_bool(result), false));
-                }
-                // These need both pointers to be in the same allocation
-                Lt | Le | Gt | Ge | Sub
-                if left_kind == right_kind
-                && (left_kind == Ptr || left_kind == usize || left_kind == isize)
-                && left.is_ptr() && right.is_ptr() => {
-                    let left = left.to_ptr()?;
-                    let right = right.to_ptr()?;
-                    if left.alloc_id == right.alloc_id {
-                        let res = match bin_op {
-                            Lt => left.offset < right.offset,
-                            Le => left.offset <= right.offset,
-                            Gt => left.offset > right.offset,
-                            Ge => left.offset >= right.offset,
-                            Sub => {
-                                return int_arithmetic!(left_kind, overflowing_sub, left.offset, right.offset);
-                            }
-                            _ => bug!("We already established it has to be one of these operators."),
-                        };
-                        return Ok((PrimVal::from_bool(res), false));
-                    } else {
-                        // Both are pointers, but from different allocations.
-                        return Err(EvalError::InvalidPointerMath);
-                    }
-                }
-                // These work if one operand is a pointer, the other an integer
-                Add | BitAnd | Sub
-                if left_kind == right_kind && (left_kind == usize || left_kind == isize)
-                && left.is_ptr() && right.is_bytes() => {
-                    // Cast to i128 is fine as we checked the kind to be ptr-sized
-                    return self.ptr_int_arithmetic(bin_op, left.to_ptr()?, right.to_bytes()? as i128, left_kind == isize);
-                }
-                Add | BitAnd
-                if left_kind == right_kind && (left_kind == usize || left_kind == isize)
-                && left.is_bytes() && right.is_ptr() => {
-                    // This is a commutative operation, just swap the operands
-                    return self.ptr_int_arithmetic(bin_op, right.to_ptr()?, left.to_bytes()? as i128, left_kind == isize);
-                }
-                _ => {}
+            if let Some(handled) = M::ptr_op(self, bin_op, left, left_ty, right, right_ty)? {
+                return Ok(handled);
```

Comment:
> I'd assume that PrimVal::to_bytes would take a Memory argument and do the logic in there.

Hm. Thinking about it, to model intptrcast faithfully we probably want to eagerly do the allocation when the cast happens, or when the transmute happens -- arithmetic operations should never have to do that, as that would mean they have side-effects and optimizations become horrible.

Anyway, we can discuss that in the intptrcast PR, whenever that happens.